File: app/src/main/java/com/arkadiusz/dayscounter/Model/RecyclerItemClickListener.java
package com.arkadiusz.dayscounter.Model;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;

/**
 * Created by Arkadiusz on 03.01.2017.
 */

public class RecyclerItemClickListener implements RecyclerView.OnItemTouchListener {

  public static interface OnItemClickListener {

    public void onItemClick(View view, int position);

    public void onItemLongClick(View view, int position);
  }

  private OnItemClickListener mListener;
  private GestureDetector mGestureDetector;

  public RecyclerItemClickListener(Context context, final RecyclerView recyclerView,
      OnItemClickListener listener) {
    mListener = listener;
    mGestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
      public boolean onSingleTapUp(MotionEvent e) {
        return true;
      }

      public void onLongPress(MotionEvent motionEvent) {
        View childView = recyclerView.findChildViewUnder(motionEvent.getX(), motionEvent.getY());
        if (childView != null && mListener != null) {
          mListener.onItemLongClick(childView, recyclerView.getChildAdapterPosition(childView));
        }

      }
    });
  }

  public boolean onInterceptTouchEvent(RecyclerView recyclerView, MotionEvent motionEvent) {
    View childView = recyclerView.findChildViewUnder(motionEvent.getX(), motionEvent.getY());
    if (childView != null && mListener != null && mGestureDetector.onTouchEvent(motionEvent)) {
      mListener.onItemClick(childView, recyclerView.getChildAdapterPosition(childView));
    }
    return false;
  }

  public void onTouchEvent(RecyclerView recyclerView, MotionEvent motionEvent) {
  }

  @Override
  public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) {
  }

}

File: app/src/main/java/com/arkadiusz/dayscounter/ui/common/RecyclerItemClickListener.kt
package com.arkadiusz.dayscounter.ui.common

import android.content.Context
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import androidx.recyclerview.widget.RecyclerView

class RecyclerItemClickListener(
    context: Context, recyclerView: RecyclerView,
    private val clickListener: OnItemClickListener
) : RecyclerView.OnItemTouchListener {

    private val mGestureDetector: GestureDetector =
        GestureDetector(context, object : GestureDetector.SimpleOnGestureListener() {
            override fun onSingleTapUp(e: MotionEvent): Boolean {
                return true
            }

            override fun onLongPress(motionEvent: MotionEvent) {
                val childView = recyclerView.findChildViewUnder(motionEvent.x, motionEvent.y)
                if (childView != null) {
                    clickListener.onItemLongClick(
                        childView,
                        recyclerView.getChildAdapterPosition(childView)
                    )
                }
            }
        })

    interface OnItemClickListener {
        fun onItemClick(view: View?, position: Int)
        fun onItemLongClick(view: View?, position: Int)
    }

    override fun onInterceptTouchEvent(
        recyclerView: RecyclerView,
        motionEvent: MotionEvent
    ): Boolean {
        val childView = recyclerView.findChildViewUnder(motionEvent.x, motionEvent.y)
        if (childView != null && mGestureDetector.onTouchEvent(motionEvent)) {
            clickListener.onItemClick(childView, recyclerView.getChildAdapterPosition(childView))
        }
        return false
    }

    override fun onTouchEvent(recyclerView: RecyclerView, motionEvent: MotionEvent) {}

    override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {}
}
